protocol/kafka_confluent/v2/protocol.go

```diff
@@ -0,0 +1,237 @@
+	}
+
+	defer in.Finish(err)
+
+	kafkaMsg := &kafka.Message{
+		TopicPartition: kafka.TopicPartition{
+			Topic:     &p.producerDefaultTopic,
+			Partition: kafka.PartitionAny,
+		},
+	}
+
+	if topic := cecontext.TopicFrom(ctx); topic != "" {
+		kafkaMsg.TopicPartition.Topic = &topic
+	}
+
+	if messageKey := MessageKeyFrom(ctx); messageKey != "" {
+		kafkaMsg.Key = []byte(messageKey)
+	}
+
+	err = WriteProducerMessage(ctx, in, kafkaMsg, transformers...)
+	if err != nil {
+		return err
+	}
+
+	err = p.producer.Produce(kafkaMsg, p.producerDeliveryChan)
+	if err != nil {
+		return err
+	}
+	e := <-p.producerDeliveryChan
+	m := e.(*kafka.Message)
+	if m.TopicPartition.Error != nil {
+		return m.TopicPartition.Error
+	}
+	return nil
+}
+
+func (p *Protocol) OpenInbound(ctx context.Context) error {
+	if p.consumer == nil {
+		return errors.New("the consumer client must be set")
+	}
+	if p.consumerTopics == nil {
+		return errors.New("the consumer topics must be set")
+	}
+
+	p.consumerMux.Lock()
+	defer p.consumerMux.Unlock()
+	logger := cecontext.LoggerFrom(ctx)
+
+	// Query committed offsets for each partition
+	if positions := TopicPartitionOffsetsFrom(ctx); positions != nil {
+		if err := p.consumer.Assign(positions); err != nil {
+			return err
+		}
+	}
+
+	logger.Infof("Subscribing to topics: %v", p.consumerTopics)
+	err := p.consumer.SubscribeTopics(p.consumerTopics, p.consumerRebalanceCb)
+	if err != nil {
+		return err
+	}
+
+	p.closerMux.Lock()
+	p.consumerCtx, p.consumerCancel = context.WithCancel(ctx)
+	defer p.consumerCancel()
+	p.closerMux.Unlock()
+
+	defer func() {
+		if !p.consumer.IsClosed() {
+			logger.Infof("Closing consumer %v", p.consumerTopics)
+			if err = p.consumer.Close(); err != nil {
+				logger.Errorf("failed to close the consumer: %v", err)
+			}
+		}
+		close(p.consumerIncoming)
+	}()
+
+	for {
+		select {
+		case <-p.consumerCtx.Done():
+			return p.consumerCtx.Err()
+		default:
+			ev := p.consumer.Poll(p.consumerPollTimeout)
+			if ev == nil {
+				continue
+			}
+			switch e := ev.(type) {
+			case *kafka.Message:
+				p.consumerIncoming <- e
+			case kafka.Error:
+				// Errors should generally be considered informational, the client will try to automatically recover.
+				// But in here, we choose to terminate the application if all brokers are down.
+				logger.Infof("Error %v: %v", e.Code(), e)
+				if p.consumerErrorHandler != nil {
+					p.consumerErrorHandler(ctx, e)
+				}
+				if e.Code() == kafka.ErrAllBrokersDown {
+					logger.Error("All broker connections are down")
+					return e
+				}
+			}
+		}
+	}
+}
+
+// Receive implements Receiver.Receive
+func (p *Protocol) Receive(ctx context.Context) (binding.Message, error) {
+	select {
+	case m, ok := <-p.consumerIncoming:
+		if !ok {
+			return nil, io.EOF
+		}
+		msg := NewMessage(m)
+		return msg, nil
+	case <-ctx.Done():
+		return nil, io.EOF
+	}
+}
+
+// Please use this method to clean up resources after utilizing the protocol
```

Comment:
Go comments must follow specific pattern

```suggestion
// Close cleans up resources after use. Must be called to properly close underlying Kafka resources and avoid resource leaks
```

Reply:
Thanks! https://tip.golang.org/doc/comment